### ISSUE TITLE
amsmath: Rationale for \AtBeginDocument{\reset@strutbox@}?

### ISSUE THREAD

**Issue by RuixiZhang42** [latex3/latex2e#207]
## Brief outline of the bug

`amsmath` does https://github.com/latex3/latex2e/blob/515099de7e392a6008280a8fb71357819ad9f923/required/amsmath/amsmath.dtx#L1531
This limits how `` can be redefined to make the strut centered around the math axis.

IMHO, the line `\AtBeginDocument{\reset@strutbox@}` is useless and/or wrong since `amsmath` already does https://github.com/latex3/latex2e/blob/515099de7e392a6008280a8fb71357819ad9f923/required/amsmath/amsmath.dtx#L1530
Useless because the resetting of `\strutbox@` at begin document will be overridden by first appearance of math mode; Wrong because `\strutbox@` is used only in math mode so users should be able to redefine it using `\fontdimen22\textfont2`.

## Minimal example showing the bug

```latex
\RequirePackage{latexbug}
\documentclass{article}
\usepackage{amsmath}

\catcode`\@=11 %

\def\reset@strutbox@{%
  \global\setbox\strutbox@\hbox{%
    \lower.5\normallineskiplimit
       \vbox{\kern-\normallineskiplimit
% Instead of copying text strut:
%        \copy\strutbox
% User can redefine math strut to center around math axis:
         \hbox{\vrule\@height\dimexpr(\baselineskip
                   +2\fontdimen22\textfont\tw@)/2\relax
                 \@depth \dimexpr(\baselineskip
                   -2\fontdimen22\textfont\tw@)/2\relax
                 \@width \z@}%
}}}

\begin{document}
Failed MWE $x$
\end{document}
```

## Log file (required) and possibly PDF file  

[amsmathstrut.log](https://github.com/latex3/latex2e/files/3800671/amsmathstrut.log)

**Comment by jfbu** [comment]
But are amsmath' `\strut@` and `\strutbox@`  destined to math mode? There is a dedicated `\Mathstrut@` and accompanying box `\Mathstrutbox@`.

**Comment by RuixiZhang42** [comment]
@jfbu Based on the `amsmath` source, `\Mathstrut@` and `\Mathstrutbox@` are only used for setting up `\big`, `\Big`, `\bigg`, `\Bigg`.

For multi-lined math objects such as matrices and cases, `amsmath` uses the box `\@arstrutbox` from the kernel. See also https://github.com/latex3/latex2e/issues/147#issuecomment-505654514

This leaves the (almost redundant) `\strut@` and `\strutbox@`, which are used in all `amsmath`’s multi-lined displays: `gathered`, `aligned[at]`, `split`, `gather[*]`, `align[at][*]`, etc.

**Comment by jfbu** [comment]
Thanks for these explanations. I should have researched myself, but now that you added them, it might be useful to others wanting to inform themselves on this issue :)

**Comment by FrankMittelbach** [comment]
> IMHO, the line `\AtBeginDocument{\reset@strutbox@}` is useless and/or wrong since `amsmath` already does
> 
> https://github.com/latex3/latex2e/blob/515099de7e392a6008280a8fb71357819ad9f923/required/amsmath/amsmath.dtx#L1530

I think you are mistaken here. \every@math@size is not called on every math formula it is only executed when when the text font size changes. So the amsmath setting is fully correct:
set the default for the main document size at document start and alter it whenever there is a \footnosize or \Large  or.

I don't consider  \reset@strutbox@ not as in any way as a macro inteded for modification from the outside, so I'm inclined to label this as wontfix/mistaken.

**Comment by RuixiZhang42** [comment]
@FrankMittelbach I disagree with
> `\every@math@size` is not called on every math formula it is only executed when when the text font size changes

I can delay my “centered around math axis” redefinition after `\AtBeginDocument{\reset@strutbox@}` to demonstrate:

```latex
\RequirePackage{latexbug}
\documentclass{article}
\usepackage{amsmath}

\catcode`\@=11 %

%%%
% Delay user redefinition after \AtBeginDocument{\reset@strutbox@}
%%%
\AtBeginDocument{%
\def\reset@strutbox@{%
  \global\setbox\strutbox@\hbox{%
    \lower.5\normallineskiplimit
       \vbox{\kern-\normallineskiplimit
% Instead of copying text strut:
%        \copy\strutbox
% User can redefine math strut to center around math axis:
         \hbox{\vrule\@height\dimexpr(\baselineskip
                   +2\fontdimen22\textfont\tw@)/2\relax
                 \@depth \dimexpr(\baselineskip
                   -2\fontdimen22\textfont\tw@)/2\relax
                 \@width \z@}%
}}}%
} % End of \AtBeginDocument

\begin{document}
Text\par
\the\ht\strutbox@\par
$x$\par
\the\ht\strutbox@
\end{document}
```

![strutchanged](https://user-images.githubusercontent.com/39968857/68213238-2a8dc200-ffa9-11e9-9709-b26fe6ccab96.png)

There are no font size changes at all. But somwhow `\reset@strutbox@` is executed twice:
- One at document begin where 70% of 12pt goes to the strut height.
- One after the inline formula `$x$` where 12pt/2+2.5pt goes to the strut height.

**Comment by FrankMittelbach** [comment]
It is executed only once at the very first formula (because then NFSS sets the math font sizes). If you uncomment the \Large it is executed again on the following formula and a third time when the group has ended on the next formula shows up.
```
\documentclass{article}
\usepackage{amsmath}

\makeatletter
\addto@hook\every@math@size{\typeout{executed \on@line}}


%%%
% Delay user redefinition after \AtBeginDocument{\reset@strutbox@}
%%%
\AtBeginDocument{%
\def\reset@strutbox@{%
                 \typeout{inside reset@strut... \on@line}%
  \global\setbox\strutbox@\hbox{%
    \lower.5\normallineskiplimit
       \vbox{\kern-\normallineskiplimit
% Instead of copying text strut:
%        \copy\strutbox
% User can redefine math strut to center around math axis:
         \hbox{\vrule\@height\dimexpr(\baselineskip
                   +2\fontdimen22\textfont\tw@)/2\relax
                 \@depth \dimexpr(\baselineskip
                   -2\fontdimen22\textfont\tw@)/2\relax
                 \@width \z@}%
}}}%
} % End of \AtBeginDocument

\begin{document}

only text

$x$

$y$

{
\Large

$x$
}

$z$

\end{document}
```

Log:
```
inside reset@strut...  on input line 32
executed  on input line 32
inside reset@strut...  on input line 39
executed  on input line 39
inside reset@strut...  on input line 42
executed  on input line 42
```

**Comment by RuixiZhang42** [comment]
@FrankMittelbach Precisely! This is the reason why I think `\AtBeginDocument{\reset@strutbox@}` is redundant and unnecessary.
- `\reset@strutbox@` globally changes `\strutbox@`.
- `\strutbox@` is used only in `amsmath`’s multi-lined displays: `gathered`, `aligned[at]`, `split`, `gather[*]`, `align[at][*]`, etc., meaning that there are *some* math modes entered so far.
- Because there are some math modes, `\reset@strutbox@` is executed at the first time such math mode is entered. This effectively overrides whatever settings done at document begin.

In summary, `\AtBeginDocument{\reset@strutbox@}` sets up a “default” strut that is *never* used in text mode. Later, whenever there is a math mode (or the first math mode after a size change), `\reset@strutbox@` is executed again.

So the effect of `\AtBeginDocument{\reset@strutbox@}` ranges only from `\begin{document}` to the first time a math formula appears, during which the internals `\strut@` and `\strutbox@` are never used. Then, in the first math formula, `\reset@strutbox@` is executed again, even if there are no size changes at all. At this point, `\fontdimen22` has become available.
